src/devices/wwan/nm-modem.c

```diff
@@ -1395,6 +1405,38 @@ nm_modem_get_data_port (NMModem *self)
 	return priv->ip_iface ?: priv->data_port;
 }
 
+int
+nm_modem_get_ip_ifindex (NMModem *self)
+{
+	NMModemPrivate *priv;
+
+	g_return_val_if_fail (NM_IS_MODEM (self), 0);
+
+	priv = NM_MODEM_GET_PRIVATE (self);
+
+	/* internally we track an unset ip_ifindex as -1.
+	 * For the caller of nm_modem_get_ip_ifindex(), this
+	 * shall be zero too. */
+	return priv->ip_ifindex != -1 ? priv->ip_ifindex : 0;
+}
+
+static void
+_set_ip_ifindex (NMModem *self, int ifindex)
+{
+	NMModemPrivate *priv = NM_MODEM_GET_PRIVATE (self);
+
+	nm_assert (ifindex >= -1);
+
+	if (priv->ip_ifindex != ifindex) {
+		gboolean notify;
+
+		notify = priv->ip_ifindex > 0;
```

Comment:
This notify behavior causes crashes when the IPv4 config comes in from the ppp-manager.  This is because the NMDevice itself is "ttyUSB0" and has no ifindex yet (because it's not a netdev).  It also has no ip_ifindex yet for the same reason.  So when the ppp-manager signals NM_PPP_MANAGER_SIGNAL_IFINDEX_SET with a valid index, in this function priv->ip_ifindex == -1, and no notification is sent.  That means the parent NMDevice never gets its ip_ifindex set.

And that makes ip4_config_merge_and_apply() create a new composite NMIP4Config without an ifindex in _ip4_config_new(), and then when we go to merge addresses to this new composite config from the WWAN one, _nm_ip_config_add_obj() hits nm_assert (ifindex > 0); and we crash.

What was the motivation for gating the notification on > 0 here?

Reply:
The idea was, that `nm_modem_get_ip_ifindex()` always returns a value >= 0. So, when setting ip-ifindex from -1 to 0, the value doesn't appear to change from outside. Such a no-change notification should be suppressed. The condition was of course wrong, and maybe the whole intent is wrong. Added a fixup to always emit the signal.

However, odd that an ignored NM_PPP_MANAGER_SIGNAL_IFINDEX_SET signal results in an assertion failure. Need to check whether that hints to another bug... (todo)

Reply:
> However, odd that an ignored NM_PPP_MANAGER_SIGNAL_IFINDEX_SET signal results in an assertion failure. Need to check whether that hints to another bug... (todo)

@thom311 if there's a bug, it would be in doing any IPv4/IPv6 operations on a device that doesn't yet have an ip_ifindex.  What I had always wanted to do but never got around to was the parent/child stuff, where the ttyUSB0 NMDevice wouldn't do anything IP-related itself, but it's child ppp0 device would, and then perhaps (for compatibility) the IP4Config from the ppp0 device would be copied to ttyUSB0 only for D-Bus compat.  Same for VPNs, PPPoE, etc.  That would have removed the whole ip_iface thing since the only IP-related operations would be performed on actual netdevices.

Obviously that would also have to be paired with API changes to expose those parent/child relationships, and not all parents would get the IPConfig copied (eg WWAN would for ppp0->ttyUSB0 but bridges wouldn't necessarily for eth0->br0).  So maybe that's confusing.  But where we have split control + data interfaces (WWAN, Bluetooth, ATM/ADSL, PPPoE, etc) it might make things simpler.  Not sure...